Use std::hypot() instead of std::sqrt(sum of squares) in computing norms of tensors.

bangerth opened the pull request:
@marcfehling 's comments in #16441 made me think that we could use `std::hypot()` also for computing norms of tensors where appropriate. This certainly leads to more accurate results, though it's not clear to me whether it leads to *faster* results. At least in principle, one could imagine that it does, though it's not clear to me that that function would be as heavily optimized as I would expect `std::sqrt()` to be.

Also somewhat related to #13321.

marcfehling commented on include/deal.II/base/tensor.h:
```diff
@@ -1720,9 +1720,30 @@ template <int rank_, int dim, typename Number>
+    }
+  if constexpr ((rank_ == 1) && (dim == 3) && std::is_arithmetic_v<Number>)
+    {
+      return std::hypot(values[0], values[1], values[3]);
```
```suggestion
      return std::hypot(values[0], values[1], values[2]);
```

bangerth replied:
Oops!

drwells commented:
I checked in with godbolt and `std::hypot()` doesn't appear to be inlined:

https://www.godbolt.org/z/4xeWMzdrz

could you check to see if there's a significant performance decrease? I don't know if this is any better or worse than calling `std::sqrt()` directly.

bangerth commented:
I don't know how to test this reliably. We can merge and see what effect that has on our performance tests, and if necessary revert?

`std::sqrt()` may have its own instruction, but I suspect that it is implementation in microcode using other instructions that also take quite a lot of cycles (see https://stackoverflow.com/questions/7724061/how-slow-how-many-cycles-is-calculating-a-square-root).

@kronbichler @peterrum You have more experience with these sorts of issues. What is your take on this?

marcfehling commented:
I should have done my research on performance before suggesting the use of `std::hypot`. I found it in the standard library and thought it would be quite fitting for some of our use cases, but all the discussion I can find deem it **safer, but slower** than a naive implementation with `std::sqrt`. We should probably stick to the faster approach.

See for example: https://stackoverflow.com/questions/32435796/when-to-use-stdhypotx-y-over-stdsqrtxx-yy

bangerth commented:
I was about to write a test such as the one in the post @marcfehling pointed to, rather than trying to benchmark a whole deal.II application. That test, on my system, shows that `std::hypot()` is about a factor of 2 (in debug mode) or 4 (in release mode) slower.

Here's the program (for reference) I'm using:
```
#include <iostream>
#include <chrono>
#include <random>
#include <algorithm>

int main(int, char**) {
  std::mt19937_64 mt;
  const auto samples = 10000000;
  std::vector<double> values(2 * samples);
  std::uniform_real_distribution<double> urd(-100.0, 100.0);
  std::generate_n(values.begin(), 2 * samples, [&]() {return urd(mt); });
  std::cout.precision(15);

  auto s1 = std::chrono::steady_clock::now();
  double sum1 = 0;
  {
    for (auto i = 0; i < 2 * samples; i += 2) {
      sum1 += std::hypot(values[i], values[i + 1]);
    }
  }
  auto e1 = std::chrono::steady_clock::now();
  auto d1 = std::chrono::duration_cast<std::chrono::microseconds>(e1 - s1);
  std::cout << std::fixed << d1.count() << "us --- s:" << sum1 << std::endl;

  auto s2 = std::chrono::steady_clock::now();
  double sum2 = 0;
  {
    for (auto i = 0; i < 2 * samples; i += 2) {
      sum2 += std::sqrt(values[i]* values[i] + values[i + 1]* values[i + 1]);
    }
  }
  auto e2 = std::chrono::steady_clock::now();
  auto d2 = std::chrono::duration_cast<std::chrono::microseconds>(e2 - s2);
  std::cout << std::fixed << d2.count() << "us --- s:" << sum2 << std::endl;

  std::cout << "Ratio: " << 1.*d1.count()/d2.count() << std::endl;
}
```
And the output:
```
tmp/g> c++ x.cc
tmp/g> ./a.out 
205549us --- s:765214218.356704235076904
102653us --- s:765214218.356704235076904
Ratio: 2.002367198230933
tmp/g> 
tmp/g> c++ x.cc -O3 -ffast-math -march=native
tmp/g> ./a.out 
20606us --- s:765214218.356554746627808
10720us --- s:765214218.356554746627808
Ratio: 1.922201492537313
```

bangerth commented:
So, in other words, we may gain a little bit of accuracy, but I don't think this is the direction we want to go into.

kronbichler commented:
Now this PR is closed (rightfully in my opinion), but let me still add a few comments: I would strongly recommend against `std::hypot` for our norm computations on tensors, because its algorithm is very conservative and taking provision for cases that are not common, and it does to so at a high cost: https://sourceware.org/git/?p=glibc.git;a=blob_plain;f=sysdeps/ieee754/dbl-64/e_hypot.c
While the benchmark does not show a large factor when compiled with `-O3 -ffast-math` (which skips most of the expensive operations in `std::hypot`), let me note that the factor is pretty big when we **do not use** `-ffast-math`, which is what our usual compilation flags do. On my machine, I see a factor of 5, which is pretty significant. Also, I note that the benchmark is accessing a large vector once, so it could be seeing a memory bandwidth limitation. I did not do the math to see how the access to main memory compares against `sqrt`, but it could make the loss with `std::hypot` look less dramatic than it is when data comes from caches, which they often do when we work with tensors.

> `std::sqrt()` may have its own instruction, but I suspect that it is implementation in microcode using other instructions 

`std::sqrt` does have its own instruction, as opposed to most other mathematical functions that go beyond `+-*/`. I also do not think `sqrt` uses microcode in modern processors. As far as I know, the state of the art is to use some radix-x hardware to perform divisions and sqrts, see http://www.imm.dtu.dk/%7Ealna/pubs/ARITH20.pdf - so there is hardware, and I would say it's almost impossible to beat with say Newton-Raphson if you aim for roundoff precision. Let me extract a few numbers from https://www.agner.org/optimize/instruction_tables.pdf using the AMD Zen 4 processor (most modern are not vastly different):
```
mulss/mulsd (multiply scalar single/double precision): latency 3 cycles, reciprocal throughput 0.5 cycles
divss (divide scalar single precision): latency 11-13 cycles, reciprocal throughput 3 cycles
divsd (divide scalar double precision): latency 11-13 cycles, reciprocal throughput 6 cycles
sqrtss (square root scalar single precision): latency 15 cycles, reciprocal throughput 5 cycles
sqrtsd (square root scalar double precision): latency 21 cycles, reciprocal throughput 8 cycles
```
In general terms, square roots are of similar costs as divisions (on Intel, the square root is even closer to division cost), and while much more expensive than multiplications/additions, much better than most other computations. In particular `std::hypot` will do some divisions (that come on top of the square root, as the hardware is shared between the two computations), comparisons, etc. - apart from `dim=1`, I see no way for `hypot` to do less operations - there certainly is no hardware for it as there is for `sqrt`.
